File: examples/undocumented/python_modular/metric_lmnn_modular.py
#!/usr/bin/env python

traindat = '../data/fm_train_real.dat'
testdat = '../data/fm_test_real.dat'
label_traindat = '../data/label_train_multiclass.dat'

parameter_list = [[traindat,testdat,label_traindat,3]]

def metric_lmnn_modular(train_fname=traindat,test_fname=testdat,label_train_fname=label_traindat,k=3):
	from modshogun import RealFeatures,MulticlassLabels,LMNN,KNN,CSVFile

	# wrap features and labels into Shogun objects
	feats_train=RealFeatures(CSVFile(train_fname))
	feats_test=RealFeatures(CSVFile(test_fname))
	labels=MulticlassLabels(CSVFile(label_train_fname))

	# LMNN
	lmnn=LMNN(feats_train,labels,k)
	lmnn.train()
	lmnn_distance=lmnn.get_distance()

	# perform classification with KNN
	knn=KNN(k,lmnn_distance,labels)
	knn.train()
	output=knn.apply(feats_test).get_labels()

	return lmnn,output

if __name__=='__main__':
	print('LMNN')
	metric_lmnn_modular(*parameter_list[0])

File: examples/undocumented/octave_modular/metric_lmnn_modular.m
modshogun;

label_train_fname='../data/label_train_multiclass.dat';
train_fname='../data/fm_train_real.dat';
test_fname='../data/fm_test_real.dat';

% wrap data into Shogun objects
feats_train=RealFeatures(CSVFile(train_fname));
feats_test=RealFeatures(CSVFile(test_fname));
labels_train=MulticlassLabels(CSVFile(label_train_fname));

% distance learning with LMNN 
disp('LMNN')

% number of target neighbours per example
k=3;
lmnn=LMNN(feats_train,labels_train,k);
init_transform=eye(feats_train.get_num_features());
lmnn.train(init_transform);

% retrieve distance learnt by LMNN
lmnn_distance=lmnn.get_distance();

% perform multiclass classification using KNN with the distance learnt
knn=KNN(k,lmnn_distance,labels_train);
knn.train();
output=knn.apply(feats_test).get_labels();
